core/e2e/tests/blockstore_server.rs

```diff
@@ -167,3 +179,120 @@ async fn e2e_blockstore_server_with_fetcher() {
+    let file3 = tokio::fs::read("tests/testdir/subdir/file3").await.unwrap();
+
+    let mut writer_sender = blockstore.file_writer().await.unwrap();
+    writer_sender.write(&file1, true).await.unwrap();
+    let root_hash_file1 = writer_sender.commit().await.unwrap();
+
+    let mut writer_sender = blockstore.file_writer().await.unwrap();
+    writer_sender.write(&file2, true).await.unwrap();
+    let root_hash_file2 = writer_sender.commit().await.unwrap();
+
+    let mut writer_sender = blockstore.file_writer().await.unwrap();
+    writer_sender.write(&file3, true).await.unwrap();
+    let root_hash_file3 = writer_sender.commit().await.unwrap();
+
+    let mut writer_sender = blockstore.dir_writer(1).await.unwrap();
+    let entry_file3 = BorrowedEntry {
+        name: "file3".as_bytes(),
+        link: BorrowedLink::Content(&root_hash_file3),
+    };
+    writer_sender.insert(entry_file3, true).await.unwrap();
+    let root_hash_subdir = writer_sender.commit().await.unwrap();
+
+    let mut writer_sender = blockstore.dir_writer(3).await.unwrap();
+    let entry_file1 = BorrowedEntry {
+        name: "file1".as_bytes(),
+        link: BorrowedLink::Content(&root_hash_file1),
+    };
+    let entry_file2 = BorrowedEntry {
+        name: "file2".as_bytes(),
+        link: BorrowedLink::Content(&root_hash_file2),
+    };
+    let entry_subdir = BorrowedEntry {
+        name: "subdir".as_bytes(),
+        link: BorrowedLink::Content(&root_hash_subdir),
+    };
+
+    writer_sender.insert(entry_file1, false).await.unwrap();
+    writer_sender.insert(entry_file2, false).await.unwrap();
+    writer_sender.insert(entry_subdir, true).await.unwrap();
+    let hash = writer_sender.commit().await.unwrap();
+
+    tokio::time::sleep(Duration::from_secs(5)).await;
+    // Fetch data from Node 1 to force getting from the other node 0
+    let fetcher = swarm.get_fetcher_socket(&pubkey2).unwrap();
+    let res = fetcher
+        .run(FetcherRequest::Fetch { hash })
+        .await
+        .expect("Failed to send request");
+
+    tokio::time::sleep(Duration::from_secs(3)).await;
+    match res {
+        FetcherResponse::Put(_) => panic!("impossible"),
+        FetcherResponse::Fetch(hash_new) => {
+            if let Err(err) = hash_new {
+                panic!("Error {err:?}");
+            }
+            let blockstore_2 = swarm.get_blockstore(&pubkey2).unwrap();
+            let dir_reader = blockstore_2
+                .get_bucket()
+                .get(&hash)
+                .await
+                .unwrap()
+                .into_dir()
+                .unwrap();
+            let mut entries = dir_reader.entries().await.unwrap();
+            let mut count = 0;
+            while let Some(Ok(entry)) = entries.next().await {
+                match entry.link {
+                    OwnedLink::Content(h) => {
+                        count += 1;
+                        assert!(blockstore_2.get_bucket().get(&h).await.is_ok());
```

Comment:
nit. maybe also compare the contents here to be super safe? either way, LGTM

Reply:
Would agree but also somewhat redundant bc of the hashing tested elsewhere, either way LGTM